### PR TITLE
Add support for Spring Boot application images built by buildpack

### PR DESCRIPTION
**Pull request by weiiwang01**

### CODE REVIEW

File: src/charm.py [hunk 1]
```diff
@@ -44,6 +44,10 @@ def _detect_java_application(self) -> ExecutableJarApplication:
                 )
             jar_file = jar_files[0]
             return ExecutableJarApplication(executable_jar_path=f"/app/{jar_file}")
+        if container.exists(
```
Comment (gregory-schiano) [inline src/charm.py:47]: I'd probably also fallback to the "buildpack" case if "/app" exists but no jar is found in the folder

File: src/java_application.py [hunk 2]
```diff
@@ -25,4 +25,25 @@ def __init__(self, executable_jar_path: str):
...
+        """Initialize the BuildpackApplication instance."""
+        self.class_path = class_path
+        self.java_executable_path = java_executable_path
+
+    def command(self) -> str:
+        """Generate the command to start the Java application in a buildpack created image."""
+        return (
+            f"{self.java_executable_path} "
+            f'-cp "{self.class_path}" '
+            f"org.springframework.boot.loader.JarLauncher"
```
Comment (jdkandersson) [inline src/java_application.py:48]: Does this need to be an f-string?

File: tests/unit/spring_boot_patch.py [hunk 3]
```diff
@@ -12,13 +13,86 @@
...
+            tar_info.size = len(content)
+            self.tar_file.addfile(tar_info, io.BytesIO(content))
+            return self
+
+        def add_dir(self, path: str) -> "OCIImageMock.OCIImageMockBuilder":
+            """Add a directory to the simulated OCI image.
+
+            Args:
+                path: the absolute path of this directory in simulated OCI image, the path should
+                 end with a slash. All parent directories will be created if not exist.
```
Comment (jdkandersson) [inline tests/unit/spring_boot_patch.py:47]: Is this indentation right?

File: tests/unit/spring_boot_patch.py [hunk 4]
```diff
@@ -49,22 +123,27 @@ def list_files(self, path: str):
             file_list.append(file_info)
         return file_list
 
-    def isdir(self, path: str):
+    def isdir(self, path: str) -> bool:
```
Comment (jdkandersson) [inline tests/unit/spring_boot_patch.py:126]: We have a new setting for `mypy` in the contributing guide that will help spot these (although not in test files): https://github.com/canonical/is-charms-contributing-guide/blob/main/CONTRIBUTING.md#type-hints

Comment (weiiwang01) [reply tests/unit/spring_boot_patch.py:126]: Added, thanks!

File: tests/unit/spring_boot_patch.py (hunk 3 above)
Comment (weiiwang01) [reply tests/unit/spring_boot_patch.py]: Fixed, thanks!

File: src/charm.py (hunk 1 above)
Comment (weiiwang01) [reply src/charm.py]: I moved the buildpack detection before executable jar detection. Now the buildpack image can be detected even if it contains the `/app` directory.

File: src/java_application.py (hunk 2 above)
Comment (weiiwang01) [reply src/java_application.py]: Removed, thanks!

**Review by gregory-schiano — APPROVED** [review]
LGTM

**Review by arturo-seijas — APPROVED** [review]
LGTM

**Review by jdkandersson — APPROVED** [review]

**Pull request merged**